### ISSUE TITLE
add `failing-issue-link` rule

### ISSUE THREAD

**Issue by jamestalmage** [avajs/eslint-plugin-ava#110]
If a user adds a failing test, they should be required to add a failing issue link.
We discussed adding that to the AVA api, but it would be just as easy to enforce that users add a comment in the vicinity of the failing test, documenting the failure.

I have some ideas for options (not all of them are necessarily good), so let's bikeshed:
## Position Option

The rule should have a `position` option, with available values being: `above | within | both`, I don't see the value of a `below` option, but I wouldn't fight it's inclusion.

``` js
// position:above 
// https://github.com/sindresorhus/eslint-plugin-ava/issues/108
test.failing(t => {});

// position: within
test.failing(t => {
  // https://github.com/sindresorhus/eslint-plugin-ava/issues/108
});
```
## Link Style Option

Allow users to configure the required GitHub link style:
- **full**: `https://github.com/sindresorhus/eslint-plugin-ava/issues/108` (`https://` should be optional)
- **short**: `sindresorhus/eslint-plugin-ava#108`
- **hash**: `#108`
- **any**: equivalent to `full|hash|short`

Full urls would always be accepted if they link outside of GH.

Short urls would be accepted in `hash` mode, if they link outside the current projects repo (which we could figure out by polling git remotes? Or just make it a config option?). 
## Issue Regexp Option

If they aren't using GitHub, then the link style option doesn't help much. Instead allow them to configure a list of regular expressions for what is allowed.

**Comment by sindresorhus** [comment]
I like the idea, but I feel it's a bit too configurable. I don't see the reason to have an option for where to place the comment. We can just say that it should be directly above the test. No practical reason to put it inside.

As for the link style, I honestly think we should just enforce a full URL. It's annoying as a contributor having to manually try to reconstruct the issue URL, and would probably deter many users. Short references like that works fine on GitHub when they're linkified, but not very user-friendly in other contexts.

However, if you strongly disagree with this, I think short/hash should be the same. I use `short` for cross-referencing issues in other repos and `hash` when it's the same repo.

**Comment by jamestalmage** [comment]
Meh. Just enforce a full valid URL, somewhere above the test. It gets the job done.

You make a good point about it not being linkified.

**Comment by jfmengels** [comment]
What happened to the `.see()` method? Pretty sure everyone gave a thumbs up for that idea :confused:

**Comment by jamestalmage** [comment]
That puts it below the test, where it's not as visible, and we would still end up writing a linter rule for it.
I doubt we want to add a `enforceSeeUseOnFailedTests` option to AVA.

**Comment by sindresorhus** [comment]
I like the comment style better to be honest. Although one benefit of the `see()` thing is that we could show the issue URL in terminal output. I guess we could still do that with some Babel helper though.

**Comment by jfmengels** [comment]
> one benefit of the see() thing is that we could show the issue URL in terminal output

Yeah, that's what I thought would happen.

> we could still do that with some Babel helper though

Sounds way more tricky to me when the user can apply any syntax in a comment.

**Comment by jamestalmage** [comment]
> Although one benefit of the see() thing is that we could show the issue URL in terminal output

Hmm. Would `iTerm2` let you click that? That would be cool. I was thinking it might just end up being more clutter. For failing tests, I was thinking the output should stay compact. No full output / error message / stack trace, just:

```
The following known issues are still failing:
  x test title one (file.js:3:12)
  x test title two (file.js:4:8)
```

**Comment by sindresorhus** [comment]
> Would iTerm2 let you click that?

Yes, just tried.

> I was thinking it might just end up being more clutter.

I think this one of the times we want clutter. People should ideally be annoyed by it and fix the issues, like with TODO warnings in ESLint. Having an easily clickable link to the issue makes it more likely some random contributor running the tests will see it in decide to fix it.

**Comment by sindresorhus** [comment]
Maybe:

```
The following known issues are still failing:
  x test title one (file.js:3:12)
    → https://github.com/sindresorhus/eslint-plugin-ava/issues/110
  x test title two (file.js:4:8)
    → https://github.com/sindresorhus/eslint-plugin-ava/issues/110
```

**Comment by jamestalmage** [comment]
I like it.

I think a Babel helper is easily doable. 
In fact, with your proposed layout, they could have multiple URLs per failing test (blocking issues, etc).

I just need a good regexp for URL scanning, and I can write the rule pretty easy

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/get-urls

**Comment by jamestalmage** [comment]
Nice! You should really just create a catalog of your modules - I miss so many trying to use npm search

**Comment by sindresorhus** [comment]
@jamestalmage Use the [GitHub search](https://github.com/search?utf8=%E2%9C%93&q=user%3Asindresorhus+url&type=Repositories&ref=searchresults) 😝

**Comment by jfmengels** [comment]
Shouldn't the discussion about whether adding the `.see()` that most agreed on, not be a discussion on the AVA repo?

**Comment by novemberborn** [comment]
@jfmengels https://github.com/sindresorhus/ava/issues/836 ;-)

**Comment by cgcgbcbc** [comment]
Does this mean forcing user put an issue link ? I find there may be a situation that people use `.failing` for represent feature for bdd/tdd, and may not have an issue for that.

**Comment by sindresorhus** [comment]
@cgcgbcbc They don't have to enable this rule. All the rules are configurable.

> may be a situation that people use .failing for represent feature for bdd/tdd

Can you elaborate? Sounds like `test.todo` or `test.skip` is what you want.

**Comment by cgcgbcbc** [comment]
> test.todo or test.skip

In my situation, I need to write test for beginners and let them implement functionality, so `todo` is not suitable for this case. I think `skip` is ok, but since it does not run the test, I should manually check If I write an already passed test(which is likely that I write a wrong test).
